clippy_lints/src/import_rename.rs

```diff
@@ -0,0 +1,88 @@
+use clippy_utils::{diagnostics::span_lint_and_sugg, source::snippet_opt};
+
+use rustc_data_structures::fx::FxHashSet;
+use rustc_errors::Applicability;
+use rustc_hir::{def::Res, Item, ItemKind, UseKind};
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_session::{declare_tool_lint, impl_lint_pass};
+use rustc_span::Symbol;
+
+declare_clippy_lint! {
+    /// **What it does:** Checks that any import matching `import_renames` is renamed.
+    ///
+    /// **Why is this bad?** This is bad for some items because their defined names are
+    /// too vauge outside of their defining scope.
+    ///
+    /// **Known problems:** Any import must use the original fully qualified path.
+    ///
+    /// For example, if you want to rename `serde_json::Value`, your clippy.toml
+    /// configuration would look like
+    /// `import-renames = [ { path: "serde_json::value::Value", rename = "JsonValue" }]` and not
+    /// `import-renames = [ { path: "serde_json::Value", rename = "JsonValue" }]` as you might expect.
+    ///
+    ///
+    /// **Example:**
+    ///
+    /// ```rust,ignore
+    /// use serde_json::Value;
+    /// ```
+    /// Use instead:
+    /// ```rust,ignore
+    /// use serde_json::Value as JsonValue;
+    /// ```
+    pub IMPORT_RENAME,
+    restriction,
+    "enforce import renames"
+}
+
+pub struct ImportRename {
+    imports: FxHashSet<(Vec<Symbol>, Symbol)>,
+}
+
+impl ImportRename {
+    pub fn new(imports: &FxHashSet<(String, String)>) -> Self {
+        Self {
+            imports: imports
+                .iter()
+                .map(|(orig, rename)| {
+                    (
+                        orig.split("::").map(|seg| Symbol::intern(seg)).collect::<Vec<_>>(),
+                        Symbol::intern(rename),
+                    )
+                })
+                .collect(),
+        }
+    }
+}
+
+impl_lint_pass!(ImportRename => [IMPORT_RENAME]);
+
+impl LateLintPass<'_> for ImportRename {
+    fn check_item(&mut self, cx: &LateContext<'_>, item: &Item<'_>) {
+        if_chain! {
+            if let ItemKind::Use(path, UseKind::Single) = &item.kind;
+            if let Res::Def(_, id) = path.res;
+            if let Some(snip) = snippet_opt(cx, item.span);
+            let use_path = cx.get_def_path(id);
+            if let Some((_, name)) = self.imports.iter().find(|(path, _)| path == &use_path);
```

Comment:
Unfortunately this will have a lot of false negatives because `get_def_path` will not be what you want if the item is re-exported. And this is fairly common.

For this to work reliably you need to resolve any paths provided to the config to a `DefId`. I think `clippy_utils::path_to_res` can help with this. Then you can compare `id` here. You can add a `check_crate` function to do the one-time work of resolving the config paths to `DefId`s.

Reply:
Sweet!! This will also fix https://github.com/rust-lang/rust-clippy/blob/master/clippy_lints/src/disallowed_method.rs#L16-L23 and `disallowed_type` making them immune to re-exports and the like.

Reply:
Good catch!

Reply:
I opened https://github.com/rust-lang/rust-clippy/pull/7345 if your curious